Enhancements for integrating RawToAces with external applications

mikaelsundell opened the pull request:
Change log:
- Implemented versioning and binary compatibility naming for libraries.
- Moved include directory to src/include for streamlined Git management.
- Added more flexible naming conventions for improved integration.
- Enhanced configuration and dependency management using find_dependencies.
- Standardized naming conventions:
  - RawToAces for classes and libraries.
  - rta for namespaces.
  - RAWTOACES for macros.
- TODO: deprecated jacobiSvd changed to colPivHouseholderQr, NOT YET VERIFIED
- TODO: evaluate performance of solver/ single/ multiple file runs

Changes to enable linking with external applications:

External applications can now link with RawToAces using the following configuration:

```
find_package (RawToAces CONFIG REQUIRED)
add_executable (${project_name} "cineontool.cpp")
target_link_libraries (${project_name}
    PRIVATE
        RawToAces::RawToAces_util2
)
```

Example C++ Code to Verify Integration:

```
#include <rawtoaces/rawtoaces_util.h>

RAWTOACES_USING_NAMESPACE

int 
main(int argc, const char * argv[])
{
    std::cout << "RawToAces: " << RAWTOACES_VERSION << std::endl;
    ImageConverter converter;
    return 0;
}
```

- Linking with RawToAces::RawToAces_util2.
- Using the standardized namespace and macros.
- Displaying the version using RAWTOACES_VERSION.

- Example from:
  https://github.com/mikaelsundell/cineontool (experimental app)

mikaelsundell committed:
Changed name of mathOps.h to mathops.h
Changed consistency across project to RawToAces
Changed jacobiSvd to colPivHouseholderQr (needs further tests and approval)
Fixed CMake indentations and formatting
Added build-* to gitignore

mikaelsundell committed:
Additional CMake naming fixes

mikaelsundell committed:
Fixed CMake indentations

mikaelsundell committed:
Fixed subdirectory paths

mikaelsundell committed:
Fixed unittests and material links in README
Fixed header mathops.h in transform_cache.cpp

mikaelsundell committed:
Fixed /W0 typo for MSVC

mikaelsundell committed:
clang-format fix

mikaelsundell committed:
Moved include to src
Replaced the cmake config files
Added find_depedency for targets
Added binary comaptibility version

mikaelsundell committed:
Added more flexible use of namespace
Added version.h.in

mikaelsundell committed:
Added OpenImageIO to find_depedency list

mikaelsundell committed:
Fixed clang-format type

antond-weta commented:
Hey Mikael , that looks good, just a few things:
- I'm not sure if we can add/modify the license on the existing files. I've only put the Apache2 license on the files I wrote from scratch.
- I would not make any changes to the solver at this stage, we may want to do that separately.
- I wonder if we could/should split your cmake fixes as a PR into master, as then rebase this branch?

mikaelsundell commented:
Hey,

> * I'm not sure if we can add/modify the license on the existing files. I've only put the Apache2 license on the files I wrote from scratch.

You're right, that might have slipped in by mistake. Should we standardize on the AMPAS license/header across all files?

```
///////////////////////////////////////////////////////////////////////////
// Copyright (c) 2013 Academy of Motion Picture Arts and Sciences
// ("A.M.P.A.S."). Portions contributed by others as indicated.
// All rights reserved.
...
```

> * I would not make any changes to the solver at this stage, we may want to do that separately.

Should we default back to jacobiSvd and make colPivHouseholderQr available as a CMake option for initial testing?

> * I wonder if we could/should split your cmake fixes as a PR into master, as then rebase this branch?

Sounds good. How do we proceed?

antond-weta commented:
> Should we standardize on the AMPAS license/header across all files?

I would leave the existing files as they are, and use Apache2 on the new files in this branch only, not on the stuff we are going to merge into master soon. Ideally I'd like to see everything relicensed as Apache2, but we'll need the code authors' permission to do that.

> Should we default back to jacobiSvd and make colPivHouseholderQr available as a CMake option for initial testing?

Yes. Also it looks like Eigen has a non-linear solver and a numerical differentiator. It would be great to see if we could get the same matrix solved with Eigen. Ceres-solver is pretty heavy and has tons of dependencies. That's on my list.

> How do we proceed?

Please do, if you don't mind. Sorry for the misdirection.

mikaelsundell commented:
> > Should we standardize on the AMPAS license/header across all files?
> 
> I would leave the existing files as they are, and use Apache2 on the new files in this branch only, not on the stuff we are going to merge into master soon. Ideally I'd like to see everything relicensed as Apache2, but we'll need the code authors' permission to do that.
> 
> > Should we default back to jacobiSvd and make colPivHouseholderQr available as a CMake option for initial testing?
> 
> Yes. Also it looks like Eigen has a non-linear solver and a numerical differentiator. It would be great to see if we could get the same matrix solved with Eigen. Ceres-solver is pretty heavy and has tons of dependencies. That's on my list.
> 
> > How do we proceed?
> 
> Please do, if you don't mind. Sorry for the misdirection.

Let me have a look at this!

mikaelsundell committed:
Changed name from rawtoaces_util.h/.cpp to image_converter.h/.cpp
Replaced type limit warnings to auto, relates to size of containers
Added default parameters for maker and model to fix segmentation fault in string compare when camera listing is requested
Added print_help when no arguments are specified
Separated parse_list_cameras and  parse_list_illuminants from parse_params to avoid exit() calls in functions

mikaelsundell committed:
Fixed clang-format

mikaelsundell committed:
Fixed clang-format

mikaelsundell committed:
Removed INSTALL_* not necessary, users can use CMAKE_INSTALL_* instead
Fixed "" issue with pkgconfig for install

mikaelsundell committed:
Moved GNUInstallDirs to the top of the file to fullt expand CMAKE_INSTALL_* variables
